### PR TITLE
Reorganize scatter arguments parsing.

### PR DESCRIPTION
**Pull request by anntzer**
I think this version is a bit easier to follow, by making the `try...
except` tighter around the parts that can actually fail, and by avoiding
the "raise a ValueError to catch it later to raise a different
ValueError" pattern.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -4224,29 +4224,33 @@ def _parse_scatter_color_args(c, edgecolors, kwargs, xsize,
...
-        # c is an array for mapping.  The potential ambiguity
-        # with a sequence of 3 or 4 numbers is resolved in
-        # favor of mapping, not rgb or rgba.
-        # Convenience vars to track shape mismatch *and* conversion failures.
+        c_is_string_or_strings = (
+            isinstance(c, str)
+            or (isinstance(c, collections.abc.Iterable) and len(c) > 0
+                and isinstance(cbook.safe_first_element(c), str)))
+
+        # After this block, c_array will be None unless c is an array for
```
Comment (timhoffm) [inline lib/matplotlib/axes/_axes.py]: Everytime I get back here, I'm wondering what `c_array` is for. It's basically a local variable in the try/else block. As stated here, the checking `c_array is None` further below carries some additional information. Wouldn't it be more clear to use a boolean like `c_is_explicit_color` or `c_should_be_mapped` here?

Or alternatively call this array `c_array_for_mapping` at least?

Comment (anntzer) [reply lib/matplotlib/axes/_axes.py]: Indeed, moving that info in a boolean variable greatly improves readability.  Pushed a separate commit with that.

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -4224,63 +4224,55 @@ def _parse_scatter_color_args(c, edgecolors, kwargs, xsize,
...
-                     len(c) > 0 and
-                     isinstance(cbook.safe_first_element(c), str))):
-            c_array = None
-        else:
+        c_is_string_or_strings = (
+            isinstance(c, str)
+            or (isinstance(c, collections.abc.Iterable) and len(c) > 0
+                and isinstance(cbook.safe_first_element(c), str)))
+
+        def invalid_shape_exception(csize, nsize):
```
Comment (efiring) [inline lib/matplotlib/axes/_axes.py]: nsize -> xsize

Comment (anntzer) [reply lib/matplotlib/axes/_axes.py]: thanks, fixed

**Pull request merged**